File: src/main/java/summarisation/SummaryFrequencyWord.java
package summarisation;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class SummaryFrequencyWord implements Comparable<SummaryFrequencyWord> {

    private String word;
    private int frequency;

    public SummaryFrequencyWord(String word, int frequency) {
        this.word = word;
        this.frequency = frequency;
    }

    /**
     * turn a map of frequency items into a list of highest ranking frequency words
     * @param map the map to process
     * @param topN how many items to return max (must be > 0)
     * @return the top N summary frequency word items in a list
     */
    public static List<String> getTopN(Map<String, Integer> map, int topN) {
        List<SummaryFrequencyWord> sfw = new ArrayList<>();
        for (String key : map.keySet()) {
            sfw.add(new SummaryFrequencyWord(key, map.get(key)));
        }
        Collections.sort(sfw);
        List<String> topWordList = new ArrayList<>();
        for (int i = 0; i < topN; i++) {
            if (i < sfw.size()) {
                topWordList.add(sfw.get(i).getWord());
            }
        }
        return topWordList;
    }

    // highest frequency first sort
    @Override
    public int compareTo(SummaryFrequencyWord o) {
        if (frequency < o.frequency) return 1;
        if (frequency > o.frequency) return -1;
        return word.compareToIgnoreCase(o.word);
    }


    public String getWord() {
        return word;
    }

    public void setWord(String word) {
        this.word = word;
    }

    public int getFrequency() {
        return frequency;
    }

    public void setFrequency(int frequency) {
        this.frequency = frequency;
    }

}

File: src/main/kotlin/summarisation/SummaryFrequencyWord.kt
package summarisation

import java.util.ArrayList

/**
 * Simple holder of a word and its frequency
 *
 */
class SummaryFrequencyWord(val word: String, val frequency: Int) : Comparable<SummaryFrequencyWord> {

    // highest frequency first sort
    override fun compareTo(other: SummaryFrequencyWord): Int {
        if (frequency < other.frequency) return 1
        return if (frequency > other.frequency) -1 else word.compareTo(other.word, ignoreCase = true)
    }

    companion object {

        /**
         * turn a map of frequency items into a list of highest ranking frequency words
         *
         * @param map the map to process
         * @param topN how many items to return max (must be > 0)
         * @return the top N summary frequency word items in a list
         */
        fun getTopN(map: HashMap<String, Int>, topN: Int): List<String> {
            val sfw = ArrayList<SummaryFrequencyWord>()
            for (key in map.keys) {
                sfw.add(SummaryFrequencyWord(key, map[key]!!))
            }
            sfw.sort()
            val topWordList = ArrayList<String>()
            for (i in 0 until topN) {
                if (i < sfw.size) {
                    topWordList.add(sfw[i].word)
                }
            }
            return topWordList
        }
    }

}

